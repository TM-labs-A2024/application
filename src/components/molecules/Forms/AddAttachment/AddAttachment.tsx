import { ArrowBackIcon, AttachmentIcon } from '@chakra-ui/icons'
import {
  Button,
  FormControl,
  FormErrorMessage,
  Input,
  Stack,
  IconButton,
  Text,
  Textarea
} from '@chakra-ui/react'
import { AttachmentFormData } from '@src/types'
import { isIOS } from '@utils/index'
import { useRouter } from 'next/navigation'
import React, { ReactElement, useMemo } from 'react'
import { useForm, FieldErrors } from 'react-hook-form'

export default function AddAttachmentForm({
  context: { onSubmit, isLoading, type, patientId, specialty }
}: {
  context: {
    onSubmit: (data: AttachmentFormData) => void
    isLoading: boolean
    type: string
    patientId: string
    specialty: string
  }
}): ReactElement {
  // --- Hooks -----------------------------------------------------------------
  const {
    handleSubmit,
    formState: { errors, isSubmitting },
    register
  } = useForm<AttachmentFormData>()

  const router = useRouter()
  // --- END: Hooks ------------------------------------------------------------

  // --- Local state -----------------------------------------------------------
  // --- END: Local state ------------------------------------------------------

  // --- Refs ------------------------------------------------------------------
  // --- END: Refs -------------------------------------------------------------

  // --- Redux -----------------------------------------------------------------
  // --- END: Redux ------------------------------------------------------------

  // --- Side effects ----------------------------------------------------------
  // --- END: Side effects -----------------------------------------------------

  // --- Data and handlers -----------------------------------------------------
  const isOrder = useMemo(() => type === 'order', [type])

  const onSubmitDetails = (data: AttachmentFormData) => {
    onSubmit(data)
  }

  const verifyErrors = (errors: FieldErrors<AttachmentFormData>) => Object.keys(errors).length > 0
  // --- END: Data and handlers ------------------------------------------------

  return (
    <div
      className={`flex h-screen w-screen flex-col overflow-hidden p-8 lg:px-96 ${isIOS() ? 'pt-20' : 'pt-8'}`}
      data-testid="add-attachment-form"
    >
      <div className="mb-8 flex flex-row items-center justify-start">
        <IconButton
          size="xl"
          className="mr-4"
          aria-label="back"
          variant="link"
          icon={<ArrowBackIcon />}
          onClick={() => {
            isOrder
              ? router.push(`/especialidad/${patientId}/${specialty}?type=order`)
              : router.push(`/especialidad/${patientId}/${specialty}?type=test`)
          }}
        />
        {isOrder && <Text className="font-medium">Nueva orden médica</Text>}
        {!isOrder && <Text className="font-medium">Nuevo análisis</Text>}
      </div>
      <form onSubmit={handleSubmit(onSubmitDetails)} className="h-full w-full">
        <FormControl
          isInvalid={verifyErrors(errors)}
          className="relative flex h-full w-full flex-col justify-between"
        >
          <Stack className="h-5/6 overflow-scroll">
            <Input
              id="title"
              placeholder={isOrder ? 'Titulo de la orden' : 'Titulo del análisis'}
              className="min-h-10"
              {...register('title', {
                required: 'Este campo es obligatorio'
              })}
            />
            <FormErrorMessage>{errors?.title && errors?.title?.message}</FormErrorMessage>
            <Textarea
              className="mt-2"
              id="description"
              placeholder="Descripción"
              {...register('description', {
                required: 'Este campo es obligatorio'
              })}
            />
            <FormErrorMessage>
              {errors?.description && errors?.description?.message}
            </FormErrorMessage>
          </Stack>
          <div className="absolute bottom-12 w-full">
            <label
              className="mb-4 inline-block h-10 min-h-10 w-full cursor-pointer rounded-md border-2 border-black p-1.5 text-center font-medium"
              htmlFor="attachment"
            >
              <AttachmentIcon /> Adjuntar imagen
            </label>
            <Input id="attachment" type="file" className="hidden" {...register('attachment')} />
            <Button isLoading={isSubmitting || isLoading} type="submit" className="w-full">
              Guardar
            </Button>
          </div>
        </FormControl>
      </form>
    </div>
  )
}
