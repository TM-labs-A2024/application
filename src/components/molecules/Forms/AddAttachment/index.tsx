import AddAttachmentForm from './AddAttachment'

export default AddAttachmentForm
