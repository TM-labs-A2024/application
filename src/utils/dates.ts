export const formatDate = (date: string) => {
  const chunks = date.split('-')

  const newDate = chunks[1] + '-' + chunks[2] + '-' + chunks[0]

  return newDate
}
