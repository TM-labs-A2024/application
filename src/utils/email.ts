import emailjs from '@emailjs/browser'

export function sendEmail(data: {
  from_name: string
  to_name: string
  code: string | number
  to_email: string
}) {
  const templateParams = {
    ...data
  }

  const serviceKey = process.env.NEXT_PUBLIC_EMAIL_SERVICE_KEY ?? ''
  const templateKey = process.env.NEXT_PUBLIC_EMAIL_TEMPLATE_KEY ?? ''
  const apiKey = process.env.NEXT_PUBLIC_EMAIL_API_KEY ?? ''

  emailjs.send(serviceKey, templateKey, templateParams, apiKey).catch(() => {
    alert('Algo sali√≥ mal, por favor reintenta el registro.')
  })
}
