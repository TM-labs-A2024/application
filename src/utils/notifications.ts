import { GENERIC_ERROR } from '@constants/index'
import { isMobile } from '@utils/index'
import { Store } from 'react-notifications-component'

export const setupErrorNotification = () => {
  Store.addNotification(
    GENERIC_ERROR('Algo salió mal, por favor inténtelo de nuevo.', isMobile(window))
  )
}
