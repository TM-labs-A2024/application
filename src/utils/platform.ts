import { Capacitor } from '@capacitor/core'

export const isWeb = () => Capacitor.getPlatform() === 'web'

export const isIOS = () => Capacitor.getPlatform() === 'ios'

export const isAndroid = () => Capacitor.getPlatform() === 'android'

export const isMobile = (window: { screen: { availWidth: number } }) =>
  Capacitor.getPlatform() === 'ios' ||
  Capacitor.getPlatform() === 'android' ||
  window?.screen?.availWidth < 425
