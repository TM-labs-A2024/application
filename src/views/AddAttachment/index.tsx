import AddAttachment from '@components/molecules/Forms/AddAttachment'
import { AttachmentFormData } from '@src/types'
import React from 'react'

export default function AddAttachmentView({
  context
}: {
  context: {
    onSubmit: (data: AttachmentFormData) => void
    isLoading: boolean
    type: string
    patientId: string
    specialty: string
  }
}) {
  return <AddAttachment context={context} />
}
