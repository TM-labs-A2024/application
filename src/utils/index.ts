import { formatDate } from './dates'
import { sendEmail, generateVerificationCode, setupEmailSending } from './email'
import { setupErrorNotification } from './notifications'
import { isWeb, isIOS, isAndroid, isMobile } from './platform'

export {
  isWeb,
  isIOS,
  isAndroid,
  isMobile,
  sendEmail,
  generateVerificationCode,
  setupEmailSending,
  setupErrorNotification,
  formatDate
}
