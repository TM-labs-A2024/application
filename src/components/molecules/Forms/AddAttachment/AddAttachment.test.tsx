import { render, screen } from '@test/utils'
import 'next/navigation'

import AddAttachment from './AddAttachment'
import '@testing-library/jest-dom'

jest.mock('next/navigation')

const props = {
  type: 'order',
  patientId: '1',
  specialty: '1'
}

const propsTest = {
  ...props,
  type: 'test'
}

describe('Molecules > Forms > AddAttachment test', () => {
  test('Depending of the type, the component shows the expected title for the orders form', () => {
    render(<AddAttachment {...props} />)

    expect(screen.getByText('Nueva orden médica')).toBeInTheDocument()
    expect(screen.queryByText('Nuevo análisis')).not.toBeInTheDocument()
  })

  test('Depending of the type, the component shows the expected title for the tests form', () => {
    render(<AddAttachment {...propsTest} />)

    expect(screen.queryByText('Nueva orden médica')).not.toBeInTheDocument()
    expect(screen.getByText('Nuevo análisis')).toBeInTheDocument()
  })

  test('Depending of the type, the component shows the placeholder text in the title input for the orders form', () => {
    render(<AddAttachment {...props} />)

    expect(screen.getByPlaceholderText('Titulo de la orden')).toBeInTheDocument()
  })

  test('Depending of the type, the component shows the placeholder text in the title input for the tests form', () => {
    render(<AddAttachment {...propsTest} />)

    expect(screen.getByPlaceholderText('Titulo del análisis')).toBeInTheDocument()
  })

  test('Matches the snapshot', () => {
    render(<AddAttachment {...props} />)

    expect(screen.getByTestId('add-attachment-form')).toMatchSnapshot()
  })
})
