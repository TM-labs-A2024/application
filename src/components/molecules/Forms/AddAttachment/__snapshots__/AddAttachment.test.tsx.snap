// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules > Forms > AddAttachment test Matches the snapshot 1`] = `
<div
  class="flex h-screen w-screen flex-col overflow-hidden p-8 lg:px-96 pt-8"
  data-testid="add-attachment-form"
>
  <div
    class="mb-8 flex flex-row items-center justify-start"
  >
    <button
      aria-label="back"
      class="chakra-button mr-4 css-84vx7q"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          fill="currentColor"
        />
      </svg>
    </button>
    <p
      class="chakra-text font-medium css-0"
    >
      Nueva orden médica
    </p>
  </div>
  <form
    class="h-full w-full"
  >
    <div
      class="chakra-form-control relative flex h-full w-full flex-col justify-between css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-stack h-5/6 overflow-scroll css-8g8ihq"
      >
        <input
          class="chakra-input min-h-10 css-1cjy4zv"
          id="title"
          name="title"
          placeholder="Titulo de la orden"
        />
        <textarea
          class="chakra-textarea mt-2 css-1msuebi"
          id="description"
          name="description"
          placeholder="Descripción"
        />
      </div>
      <div
        class="absolute bottom-12 w-full"
      >
        <label
          class="mb-4 inline-block h-10 min-h-10 w-full cursor-pointer rounded-md border-2 border-black p-1.5 text-center font-medium"
          for="attachment"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21.843,3.455a6.961,6.961,0,0,0-9.846,0L1.619,13.832a5.128,5.128,0,0,0,7.252,7.252L17.3,12.653A3.293,3.293,0,1,0,12.646,8L7.457,13.184A1,1,0,1,0,8.871,14.6L14.06,9.409a1.294,1.294,0,0,1,1.829,1.83L7.457,19.67a3.128,3.128,0,0,1-4.424-4.424L13.411,4.869a4.962,4.962,0,1,1,7.018,7.018L12.646,19.67a1,1,0,1,0,1.414,1.414L21.843,13.3a6.96,6.96,0,0,0,0-9.846Z"
              fill="currentColor"
            />
          </svg>
           Adjuntar imagen
        </label>
        <input
          class="chakra-input hidden css-1cjy4zv"
          id="attachment"
          name="attachment"
          type="file"
        />
        <button
          class="chakra-button w-full css-ez23ye"
          type="submit"
        >
          Guardar
        </button>
      </div>
    </div>
  </form>
</div>
`
